utils/job_db.py

```diff
@@ -17,6 +17,26 @@ def create_index():
                           ("user", DESCENDING)])
 
 
+def generate_unique_object_identifier():
+    """
+    Create the unique object identifier.
+
+    This is NOT the whole object id, just the last part to ensure every object
+    is unique and the last characters are digits.
+    :return:
+    """
+    new_id = db.objects.find_one_and_update(
+        {'next_object_id': {'$exists': True}},
+        {'$inc': {'next_object_id': 1}},
+        return_document=ReturnDocument.AFTER
+    )
+    if not new_id:
+        first_object_id = int(os.environ['FIRST_OBJECT_ID'])
+        db.objects.insert_one({'next_object_id': first_object_id})
+        return first_object_id
+    return new_id['next_object_id']
```

Comment:
Kann es so theoretisch nicht passieren, dass die erste ID mehrfach vergeben wird, wenn mehrere Prozesse gleichzeitig laufen?

Könnte man die erste ID beim Starten von Cilantro in die DB schreiben um das zu verhindern?

Reply:
Kann ich machen. Haben wir einen enstprechenden on_start hook bzw eine Methode, die dann ausgeführt wird?

Reply:
So einen hook gibt es noch nicht. Vielleicht geht es auch einfacher, indem man ein `find_one_and_update` mit dem Filter `{'$exists': False}` voranschickt, der das Dokument mit der id erstellt?

Reply:
Das funktioniert mit find_one_and_update leider nicht.

$exists: False funktioniert so mMn nicht. Wenn es kein Dokument mit dem Key 'next_object_id' gibt, gibt es auch nichts zurück. Es würde nur ein Dokument zurückgegeben werden, wenn der key 'next_object_id' existiert, aber kein Value hat.

Man muss also insert_one nutzen. Leider ist mir schleierhaft, warum insert_one über kein Write Concern verfügt.

Was ich stattdessen machen könnte, ist vor oder nach dem Erstellen des Index zu prüfen ob es 'next_object_id' gibt und es ggf. hinzuzufügen. In der Hoffnung, dass da noch keine Parallelisierungsprobleme auftreten. Schön ist das aber auch nicht.

Reply:
Eigentlich müsste es mit dem $setOnInsert-Operator gehen, siehe:
* https://stackoverflow.com/a/16362833/1245675

Oder?

Reply:
Leider nicht. SetOnInsert und inc sind nicht kompatibel. Wenn die auf das selbe field geht, wirft mongoDB einen Fehler 
```pymongo.errors.OperationFailure: Updating the path 'next_object_id' would create a conflict at 'next_object_id'
```
Hier ist das gleiche Problem und ein Lösungsvorschlag https://stackoverflow.com/a/40770440

Reply:
Ja stimmt, [dieser Kommenatr](https://stackoverflow.com/questions/40769583/inc-vs-setoninsert-in-mongocollection-findoneandupdate/40770440#comment68994295_40770440) sagt ja im Prinzip da gleiche wie ich. Vertracktes Problem. Dann müssen wir doch irgendwie sicherstellen, dass die initial object_id beim Starten in die db geschrieben wird.

Vielleicht hilft uns der Mechanismus, der [hier](https://hub.docker.com/_/mongo) unter "Initializing a fresh instance" beschrieben ist?